### PR TITLE
Deprecate `get_many` method name in favor of `get_multi` to conform with expected memcached API

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR will deprecate the existing `get_many` method in favor of `get_multi`.  `get_many` now exists as an alias for `get_multi`, and will be fully removed in a future version.

### Details - How are you making this change?  What are the effects of this change?

closes: https://github.com/Shopify/caching-platform/issues/721

This PR had originally performed additional refactoring to attempt to return a `HashMap<Result<Value, Error>, Error>`, but it's not possible to efficiently map errors on a per-key basis at this level due to how the memcached server returns errors.  If there is an error would be returned from the memcached server, then no `Value`s are returned, even for properly formed keys that would result in a cache hit.  These potential errors must be handled by the client.  As such, the crate method should be left as barebones as possible for performance reasons, so the `Vec<Value>` returned from `get_multi` makes the most sense.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -178,10 +185,46 @@ impl Client {
...
 
         match self.get_read_write_response().await? {
-            Response::Status(s) => Err(s.into()),
-            Response::Data(d) => d.ok_or(Status::NotFound.into()),
-            _ => Err(Status::Error(ErrorKind::Protocol(None)).into()),
+            Response::Status(s) => return Err(s.into()),
+            Response::Data(Some(items)) => {
+                let mut map: HashMap<Vec<u8>, Value> = items
+                    .into_iter()
+                    .map(|item| (item.key.clone(), item))
```
Comment (mrattle) [inline src/lib.rs:192]: This initial `.map()` is required for ownership reasons - we initialize a hashmap that contains a the keys and then `.remove()` them into the results hashmap as we go.

This seemed to be the natural path away form using something like
```
for item in items {
                    if let Some(key) = keys.iter().find(|k| k.as_ref() == item.key) {
                        results.insert(key.clone(), Ok(item));
                    }
                }
```

File: src/lib.rs [hunk 2]
```diff
@@ -163,12 +163,19 @@ impl Client {
     /// describes the metadata and data of the key.
     ///
     /// Otherwise, [`Error`] is returned.
-    /// This will eventually be deprecated in favor of `get_multi`
-    pub async fn get_many<I, K>(&mut self, keys: I) -> Result<Vec<Value>, Error>
+    pub async fn get_multi<'a, K>(
+        &mut self,
+        keys: &'a [K],
+    ) -> Result<HashMap<&'a K, Result<Option<Value>, Error>>, Error>
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/lib.rs]: `<&[u8]>` doesn't seem to cooperate, but changing to `Vec<u8>` was trivial and simplified the code significantly.

File: src/lib.rs [hunk 3]
```diff
@@ -163,12 +164,19 @@ impl Client {
     /// describes the metadata and data of the key.
     ///
     /// Otherwise, [`Error`] is returned.
-    /// This will eventually be deprecated in favor of `get_multi`
-    pub async fn get_many<I, K>(&mut self, keys: I) -> Result<Vec<Value>, Error>
+    pub async fn get_multi<'a, K>(
+        &mut self,
+        keys: &'a [K],
```
Comment (nickamorim) [inline src/lib.rs:169]: If we change this to a `&[u8]` does it remove the need to clone the key?

File: src/lib.rs [hunk 4]
```diff
@@ -163,12 +164,19 @@ impl Client {
...
     where
-        I: IntoIterator<Item = K>,
-        K: AsRef<[u8]>,
+        K: AsRef<[u8]> + Eq + std::hash::Hash + Clone,
     {
+        if keys.is_empty() {
+            return Ok(FxHashMap::default());
+        }
+
+        let mut results = FxHashMap::with_capacity_and_hasher(keys.len(), Default::default());
```
Comment (nickamorim) [inline src/lib.rs:178]: What's `Default::default()`?

File: tests/integration_tests.rs [hunk 5]
```diff
@@ -272,16 +272,48 @@ async fn test_get_many_with_nonexistent_key() {
         );
     }
 
-    keys.push("thisisakeythatisntset");
+    let unset_key = "thisisakeythatisntset";
```
Comment (nickamorim) [inline tests/integration_tests.rs:275]:
```suggestion
    let unset_key = "thisisakeythatisnotset";
```

File: tests/integration_tests.rs [hunk 6]
```diff
@@ -272,16 +272,48 @@ async fn test_get_many_with_nonexistent_key() {
...
 
     assert_eq!(original_keys_length, results.len());
 
+    // Hashmap should only contain keys that have a cache hit
+    assert!(!results.contains_key(unset_key.as_bytes()));
+
     for result in results {
-        let key_str = std::str::from_utf8(&result.key).unwrap();
-        assert!(keys.clone().contains(&key_str));
+        if let Some(value) = result.1.expect("should have unwrapped a Result") {
```
Comment (nickamorim) [inline tests/integration_tests.rs:287]: Isn't the expect msg more like "value to be present" or something?

File: src/lib.rs (hunk 3 above)
Comment (mrattle) [reply src/lib.rs:169]: Nick and I went through a quick exercise trying to get this suggestion to work and it seems like we always require either multiple iterations or a secondary hashmap that contains the keys, which is almost certainly more expensive than just cloning the key as we insert into the results hashmap.  The `Vec<u8>` approach feels like the best compromise.

File: src/lib.rs [hunk 7]
```diff
@@ -178,10 +186,33 @@ impl Client {
...
-            Response::Status(s) => Err(s.into()),
-            Response::Data(d) => d.ok_or(Status::NotFound.into()),
-            _ => Err(Status::Error(ErrorKind::Protocol(None)).into()),
+            Response::Status(s) => return Err(s.into()),
+            Response::Data(Some(items)) => {
+                for item in items {
+                    results.insert(item.key.clone(), Ok(Some(item)));
+                }
+            }
+            Response::Data(None) => {}
```
Comment (drinkbeer) [inline src/lib.rs:195]: If there is no data, are we just returning en empty HashMap?

File: src/lib.rs [hunk 8]
```diff
@@ -178,10 +186,33 @@ impl Client {
...
 
         match self.get_read_write_response().await? {
-            Response::Status(s) => Err(s.into()),
-            Response::Data(d) => d.ok_or(Status::NotFound.into()),
-            _ => Err(Status::Error(ErrorKind::Protocol(None)).into()),
+            Response::Status(s) => return Err(s.into()),
+            Response::Data(Some(items)) => {
+                for item in items {
+                    results.insert(item.key.clone(), Ok(Some(item)));
+                }
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/lib.rs]: This would result in a hashmap containing keys for which there is a cache hit and the corresponding Value payload received from the server, or a single error.  There is no way to map errors at a per-key level, as the server either will not respond in the case of a cache miss or will respond with a single error for the whole request command if there is a malformed key or something.  For this reason, it's not worth doing the work to transform the result set into a hashmap, and it's better to leave the function signature as-is.  By mapping to a hashmap, we are always forcing an iteration over the vector of results, effectively for no gain.  It's better to have the client do that work, and probably doesn't actually introduce any additional overhead at that layer since it's likely that the client would have to iterate over the list of keys that was passed in to the `get_multi` method in the first place.

**Review by nickamorim — APPROVED** [review]
